lib/rspec-console/config_cache.rb

```diff
@@ -1,82 +1,84 @@
-  # RSpec.configuration gets also loaded with a bunch of stuff from the
-  # 'spec/spec_helper.rb' file. Often that instance is extended with other
-  # modules (FactoryGirl, Mocha,...) and we don't want to replace requires with
-  # load all around the place.
-  #
-  # Instead, we cache whatever is done to RSpec.configuration during the
-  # first invokration of require('spec_helper').
-  # This is done by interposing the Proxy class on top of RSpec.configuration.
-  #
-  attr_accessor :proxy, :recorded_config, :shared_examples_groups
-
-  def initialize
-    ::RSpec.instance_eval do
-      def self.configuration=(value)
-        @configuration = value
-      end
-    end
-  end
+  attr_accessor :proxy, :recorded_registry
 
   def cache
-    if self.proxy
-      # replay
-      ::RSpec.configure do |config|
-        self.recorded_config.each do |msg|
-          config.send(msg[:method], *msg[:args], &msg[:block])
+    if have_recording?
+      ::RSpec.configure &replay_recorded_config
+
+      # RSpec 2 and 3 have different APIs for accessing shared_examples. 3 has
+      # the concept of a "registry" whereas 2 does not.
+      if version >= Gem::Version.new('3')
+        unless shared_examples.empty?
+          ::RSpec.world.
+            shared_example_group_registry.
+          add(
+            :main,
+            shared_examples.keys.first,
+            &shared_examples.values.first
+          )
         end
+      else
+        ::RSpec.world.
+          shared_example_groups.merge!(recorded_registry || {}) rescue nil
       end
-      ::RSpec.world.shared_example_groups.merge!(self.shared_examples_groups || {}) rescue nil
-
     else
       # record
-      real_config = ::RSpec.configuration
-      self.recorded_config = []
-      self.proxy = Proxy.new(self.recorded_config, real_config)
+      original_config = ::RSpec.configuration
+      self.proxy = RSpecConsole::Proxy.new(original_config)
+
       ::RSpec.configuration = self.proxy
-      yield
-      ::RSpec.configuration = real_config
-      self.shared_examples_groups = ::RSpec.world.shared_example_groups.dup rescue nil
 
-      # rspec-rails/lib/rspec/rails/view_rendering.rb add methods on the
-      # configuration singleton. Need advice to copy them without going down
-      # the road with object2module.
-    end
+      yield # spec helper is called during this yield, see #reset
 
-    # Well, instead of copying them, we redirect them to the configuration
-    # proxy. Looks like it good enough.
-    proxy = self.proxy
-    ::RSpec.configuration.singleton_class.send(:define_method, :method_missing) do |method, *args, &block|
-      proxy.send(method, *args, &block)
-    end
+      ::RSpec.configuration = original_config
+
+      if version >= Gem::Version.new('3')
+        self.recorded_registry = ::RSpec.world.
+          shared_example_group_registry.dup rescue nil
+      else
+        self.recorded_registry = ::RSpec.world.
+          shared_example_groups.dup rescue nil
```

Comment:
i like to respect a line length limit. in fact, for readability. ruby style guide accepts trailing dot or dot on new line, i prefer trailing dot.